TF-3601 improvement: add rule to check for correct decorator order

joecan opened the pull request:
##### *To:*
@jjinux @inancsevinc @kevinzang @cansin 

##### *What:*
`udemy/decorator-order` rule checks the order of `@autobind @action` on
method definitions and `@inject @observer` on class declarations.

##### *JIRA:*
https://udemyjira.atlassian.net/browse/TF-3601

##### *What did you test:*
Just the included tests.js.

##### *What dashboards will you be monitoring:*
N/A

joecan committed:
improvement: add rule to check for correct decorator order

kevinzang commented on packages/eslint-plugin-udemy/rules/decorator-order/index.js:
```diff
@@ -0,0 +1,29 @@
+        create(context) {
+            return {
+                MethodDefinition: orderChecker(context, 'autobind', 'action'),
+                ClassDeclaration: orderChecker(context, 'inject', 'observer'),
```
Should the logic be `@observer should be the first decorator`, rather than `@observer should be before @inject`?

I thought in general `@observer` should be first: https://udemy.slack.com/archives/C04HUGFL0/p1530223702000052?thread_ts=1530219678.000357&cid=C04HUGFL0

There could be other decorators besides `@inject`. For example, in the Slack thread, Carmen was using `@deviceType`.

joecan replied:
This can also get confusing as to what we mean by "first".  In the AST there is simply an array of decorators ordered how they are written left to right. I think by "first" you mean inner-most, which would be last in the array. 

I was writing the rule with another thread in mind where people were mostly concerned with the order of decorators relative to one another. I will change my check to make sure that if `@observer` is there that it is last in the array. But does the same go for `@action` or is it that `@autobind` needs to go first, or both?

kevinzang replied:
My bad- yes, by "first" I mean "inner-most".

As far as I'm aware, the only problem we've seen regarding the `@action` decorator is that is that `@action @autobind` doesn't work, so I think what you currently have is fine. I think enforcing `@action` to be innermost is probably fine as well, but I'm not 100% sure about it.

kevinzang reviewed (approved):
@joecan LGTM!

joecan committed:
fix: checking that @observer is the inner-most decorator

joecan committed:
Publish

 - babel-polyfill-udemy-website@7.0.6
 - babel-preset-udemy-website@8.0.6
 - eslint-config-udemy-basics@5.0.2
 - eslint-config-udemy-website@7.1.4
 - eslint-plugin-udemy@6.0.2

jjinux commented on packages/eslint-plugin-udemy/rules/decorator-order/README.md:
````diff
@@ -0,0 +1,28 @@
+The following patterns are considered warnings:
+
+```js
+
````
Do you need these blank lines around the code?

jjinux commented on packages/eslint-plugin-udemy/rules/decorator-order/index.js:
```diff
@@ -0,0 +1,37 @@
+                MethodDefinition(node) {
+                    const decorators = node.decorators;
+                    if (!decorators) { return; }
+                    const decoratorNames = decorators.map(d => d.expression.name);
```
I wonder if this will fail if the decorator is a function call.

jjinux commented on packages/eslint-plugin-udemy/rules/decorator-order/tests.js:
```diff
@@ -0,0 +1,52 @@
+        },
+        {
+            code: `@observer @inject class MyClass {
+                      @autobind @action mymethod() { return true; }
```
nit: indentation

jjinux reviewed (commented):
I left a few comments. LGTM!
